pkg/pagination/pagination.go

```diff
@@ -0,0 +1,103 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package pagination implements a series of utilities for dealing with
+// paginating lists of objects for an API.
+package pagination
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/fernet/fernet-go"
+)
+
+// ErrInvalidToken is returned when a token fails to Unmarshal because it was
+// invalid or expired.
+var ErrInvalidToken = errors.New("invalid or expired pagination token")
+
+// Key represents the key used to cryptographically secure the token
+// being used to keep track of pages.
+type Key string
+
+// MustGenerateNewKey generates a new pagination key.
+//
+// This function will panic if your system does not have enough entropy to
+// produce the required randomness.
+func MustGenerateNewKey() Key {
+	var key fernet.Key
+	if err := key.Generate(); err != nil {
+		panic("pagination: failed to read enough randomness to generate a new key")
+	}
+
+	return Key(key.Encode())
+}
+
+// KeyFromString creates the key for a given string.
+//
+// Strings must be 32-bit URL-safe base64 representations of the key bytes.
+func KeyFromString(keyString string) (k Key, err error) {
+	var key *fernet.Key
+	key, err = fernet.DecodeKey(keyString)
+	if err != nil {
+		err = errors.New("invalid pagination key string: must be 32-bit URL-safe base64")
+	}
+	return Key(key.Encode()), err
+}
+
+// MustKeyFromString creates the key for a given string.
+//
+// This function will panic if the key is not well-formed.
+//
+// You may find this function useful if you have already called KeyFromString
+// and know the provided value is safe.
+func MustKeyFromString(keyString string) Key {
+	k, err := KeyFromString(keyString)
+	if err != nil {
+		panic("pagination: failed to encode string as key")
+	}
+
+	return k
+}
+
+// String implements the fmt.Stringer interface for Key.
+func (k Key) String() string {
+	return string(k)
+}
+
+// MarshalToken encodes an interface into JSON bytes and encrypts it.
+func (k Key) MarshalToken(v interface{}) ([]byte, error) {
+	var buf bytes.Buffer
+	err := json.NewEncoder(&buf).Encode(v)
+	if err != nil {
+		return nil, err
+	}
+
+	decoded, _ := fernet.DecodeKey(string(k))
+	return fernet.EncryptAndSign(buf.Bytes(), decoded)
+}
+
+// UnmarshalToken decrypts a token using provided key and decodes the result
+// into the provided interface.
+func (k Key) UnmarshalToken(token string, v interface{}) error {
+	decoded, _ := fernet.DecodeKey(string(k))
```

Comment:
Do we really want to suppress the error here? Yes, it was not handled before. If there's a valid reason, I think it makes sense to comment it.